tests/unit/db/query_result_tests.py

```diff
@@ -0,0 +1,380 @@
+
+    def test_iteration_with_skip_option_fails(self):
+        """
+        Test __iter__() fails when "skip" option is provided
+        """
+        self.populate_db_with_documents(100)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            skip=10
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(
+                str(err),
+                'Cannot use skip for iteration'
+            )
+
+    def test_iteration_with_limit_option_fails(self):
+        """
+        Test __iter__() fails when "limit" option is provided
+        """
+        self.populate_db_with_documents(100)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            limit=10
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(
+                str(err),
+                'Cannot use limit for iteration'
+            )
+
+    def test_iteration_invalid_page_size(self):
+        """
+        Test __iter__() fails as expected when page_size is set to 0 or less
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=0
+        )
+        try:
+            for doc in result:
+                self.fail('Above statement should raise an Exception')
+        except CloudantArgumentError, err:
+            self.assertEqual(str(err), 'Invalid page_size: 0')
+
+    def test_iteration_result_eq_page_size(self):
+        """
+        Test __iter__() works as expected when the result size is equal to the
+        page_size
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=5
+        )
+        self.assertEqual(
+            [doc['_id'] for doc in result],
+            ['julia000', 'julia001', 'julia002', 'julia003', 'julia004']
+        )
+
+    def test_iteration_result_gt_page_size(self):
+        """
+        Test __iter__() works as expected when the result size is more than the
+        page_size
+        """
+        self.populate_db_with_documents(5)
+        result = QueryResult(
+            Query(self.db),
+            selector={'_id': {'$gt': 0}},
+            fields=['_id', '_rev'],
+            page_size=3
+        )
+        self.assertEqual(
+            [doc['_id'] for doc in result],
+            ['julia000', 'julia001', 'julia002', 'julia003', 'julia004']
```

Comment:
I think what might be the problem is the name of the `page_size` parameter.  Its purpose is to dictate the batch size for Result iterations.  Perhaps a better name would be `batch_size`.  We use skip and limit in the iterator logic along with the page_size to manage the stream of documents.  This is the reason why we don't allow skip and limit parameters on iterations of Results/QueryResults.  So in this example the total number of documents matching the query is 5 but the page_size is set to 3 which tells the iterator to retrieve batches of at most 3 documents at a time until you've retrieved all of the documents matching the query.